### PR TITLE
remake GTRecipeWidget add oc button

### PR DESCRIPTION
**Pull request by Arborsm**
## What
feat: add oc button
fixed: some recipe too short to display

## Implementation Details
remake `GTRecipeWidget`

## Outcome
screenshot

## Potential Compatibility Issues
some addon reicpe might overlap with oc button

### CODE REVIEW

File: src/main/java/com/gregtechceu/gtceu/integration/GTRecipeWidget.java [hunk 1]
```diff
@@ -194,34 +192,148 @@ public GTRecipeWidget(GTRecipe recipe) {
...
+            LabelWidget labelWidget = new LabelWidget(3 - xOffset, textsY, text).setTextColor(-1).setDropShadow(true);
+            addWidget(labelWidget);
+            recipeParaTexts.add(labelWidget);
+        }
+        if (inputEUt != 0) {
+            LabelWidget voltageTextWidget = new LabelWidget(getVoltageXOffset() - xOffset, getSize().height - 10, tierText).setTextColor(-1).setDropShadow(false);
+            if (recipe.recipeType == GTRecipeTypes.FUSION_RECIPES || recipe.recipeType == GTRecipeTypes.GAS_COLLECTOR_RECIPES) {
+                voltageTextWidget.setSelfPositionY(getSize().height - 20);
+            }
+            // make it clickable
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/integration/GTRecipeWidget.java]: Why is this hardcoded? It should be some parameter in GTRecipeTypeUI imo.

File: src/main/java/com/gregtechceu/gtceu/integration/GTRecipeWidget.java [hunk 2]
```diff
@@ -194,34 +192,148 @@ public GTRecipeWidget(GTRecipe recipe) {
...
+        }
+        setRecipeTextWidget();
+    }
+
+    private void setRecipeTextWidget() {
+        long inputEUt = RecipeHelper.getInputEUt(recipe);
+        int duration = recipe.duration;
+        String tierText = GTValues.VNF[tier];
+        if (tier > getMinTier() && inputEUt != 0) {
+            LongIntPair pair = OverclockingLogic.NON_PERFECT_OVERCLOCK.getLogic().runOverclockingLogic(
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/integration/GTRecipeWidget.java]: Why is this hardcoded to non-perfect OC? make it changeable via (maybe the UI? Maybe the recipe type?)

**Review by screret — CHANGES_REQUESTED** [review]

File: src/main/java/com/gregtechceu/gtceu/integration/GTRecipeWidget.java (hunk 2 above)
Comment (Arborsm) [reply src/main/java/com/gregtechceu/gtceu/integration/GTRecipeWidget.java]: oh forgot

File: src/main/java/com/gregtechceu/gtceu/api/recipe/GTRecipeType.java [hunk 3]
```diff
@@ -84,6 +83,12 @@ public class GTRecipeType implements RecipeType<GTRecipe> {
     private CompoundTag customUICache;
     @Getter
     private final GTRecipeLookup lookup = new GTRecipeLookup(this);
+    @Setter
+    @Getter
+    private boolean upwardVoltageText = false;
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/api/recipe/GTRecipeType.java]: maybe call it `offsetVoltageText`?

File: src/main/java/com/gregtechceu/gtceu/api/recipe/GTRecipeType.java [hunk 4]
```diff
@@ -84,6 +83,12 @@ public class GTRecipeType implements RecipeType<GTRecipe> {
...
+    @Setter
+    @Getter
+    private int upwardVoltageTextHeight = 20;
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/api/recipe/GTRecipeType.java]: `voltageTextOffset`

**Review by screret — CHANGES_REQUESTED** [review]

**Comment by Arborsm** [comment]
done

**Review by screret — APPROVED** [review]

**Comment by AzakoYin** [comment]
?

Uploading QQ202442-231816.mp4…

**Comment by AzakoYin** [comment]
https://github.com/GregTechCEu/GregTech-Modern/assets/123106150/a023150b-b59c-4eec-9aad-958546abfd89

version: 1.1.4b build569

**Comment by AzakoYin** [comment]
{2377E481-4C7E-4312-9B4E-62BFB91B39B3}
???
Version  1.1.4b build 569      lang : en_us

**Comment by Arborsm** [comment]
> {2377E481-4C7E-4312-9B4E-62BFB91B39B3}
> ???
> Version  1.1.4b build 569      lang : en_us

cause screret make some change it will fix by another pr

**Comment by AzakoYin** [comment]
> > {2377E481-4C7E-4312-9B4E-62BFB91B39B3}
> > ???
> > Version  1.1.4b build 569      lang : en_us
> 
> cause screret make some change it will fix by another pr

got it

**Pull request merged**